## Reduction operation: the sum of the numbers in the range [0, value)

In [3]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.035003662109375 seconds
35.8 ms ± 173 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



a)En la siguiente celda de código del notebook vamos a utilizar el tipo de datos listas para realizar
dicha operación de reducción. Para ello, crea una lista de Python con la función range que contenga
10^6 elementos. Una vez creada, calcula el tiempo que tardas en sumar todos los elementos de la lista
de 2 formas diferentes: a) Usando un bucle for similar al código ya mostrado en la celda anterior
para ir sumando al resultado total elemento a elemento de la lista, y b) usando la función sum que
tienen las listas. Como verás, hay una diferencia importante a favor de usar la función sum.

In [2]:

N = 10**6
lista_datos = list(range(N))

# Funcion para bucle for
def suma_manual(datos):
    "Suma los elementos de una lista usando un bucle for"
    total = 0
    for elemento in datos:
        total += elemento
    return total

tiempo_bucle = %timeit -r 2 -o -q suma_manual(lista_datos)

print("Tiempo tomado usando bucle for:", tiempo_bucle)
print(f"Resultado de la suma: {suma_manual(lista_datos)}\n")


# Metodo B
tiempo_sum = %timeit -r 2 -o -q sum(lista_datos)

print("Tiempo tomado usando sum():", tiempo_sum)
print(f"Resultado de la suma: {sum(lista_datos)}")

Tiempo tomado usando bucle for: 26.1 ms ± 79.6 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)
Resultado de la suma: 499999500000

Tiempo tomado usando sum(): 5.82 ms ± 6.98 μs per loop (mean ± std. dev. of 2 runs, 100 loops each)
Resultado de la suma: 499999500000


In [ ]:
b)En la segunda celda de código vamos a utilizar el tipo de datos array que nos proporciona el paquete
numpy. En primer lugar, pasa la lista creada en la celda anterior a un array de numpy. Y al igual que
antes, calcula el tiempo que tardas en sumar todos los elementos del array de 2 formas diferentes: a)
Usando un bucle for para ir sumando elemento a elemento del array al resultado total, y b) usando
la función sum que tiene el paquete numpy para los arrays.

In [2]:
import numpy as np
N = 10**6
lista_datos = list(range(N))
array_datos = np.array(lista_datos)

#Funcion para sumar en bucle
def suma_manual_numpy(arr):
    s = 0
    for i in arr:
        s += i
    return s

# Usando bucle for sobre el Array .Apartado a)
tiempo_bucle_np = %timeit -r 2 -o -q suma_manual_numpy(array_datos)

print("Tiempo tomado usando bucle for en NumPy:", tiempo_bucle_np)
print(f"Resultado de la suma: {suma_manual_numpy(array_datos)}\n")


# Usando sum. Apartado b)
tiempo_numpy_sum = %timeit -r 2 -o -q np.sum(array_datos)

print("Tiempo tomado usando np.sum():", tiempo_numpy_sum)
print(f"Resultado de la suma: {np.sum(array_datos)}")

Tiempo tomado usando bucle for en NumPy: 52.4 ms ± 412 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)
Resultado de la suma: 499999500000

Tiempo tomado usando np.sum(): 231 μs ± 251 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)
Resultado de la suma: 499999500000


c) Crea una nueva celda de texto debajo de la última celda de código para explicar los resultados
obtenidos por los 3 procedimientos (código original, listas y numpy).

El código original empleando un bucle for tarda aproximadamente 35.8 ms , este método es el que aplica directamente fuerza bruta en la suma, sin aprovechar ninguna estructura de datos.

En el caso de emplear listas mediante python en la suma, este proceso es más rápido pero en el bucle for python necesita comporbar que cada elemento de la lista cumple los requisitos para ser sumado antes de continuar. Esto hace que aun siendo más efectivo presente cierto retraso en el proceso en comparación con el comando sum(), este comando se implementa en lenguaje C por lo que es capaz de iterar los elementos de la lista a mayor velocidad. Esto hace que los dos métodos empleados en el apartado a) tarden 26.1 ms y 5.82 ms respectivamente.

Por último, empleando la arquitectura de arrays de numpy (apartadob)) en la que vemos notorias diferencias entre emplear el bucle for y el comando sum(). Aunque es la estructura más eficiente de almacenamiento de datos, al emplear el bucle for tenemos que acceder a cada elemento del array individualmente y si fuera necesario comprobar y cambiar su formato para poder sumarlo. Esto hace que el proceso llegue a ralentizarse tardando 52.4 ms. Por otra parte, al emplear arrays y el comando sum() nos permite aprovechar la vectorización del array y se emplean en el proceso de suma funciones en las que el procesador está altamente optimizado, llegando a tardar solamente 231 μs.El procesador trabaja más eficientemente mediante arrays de numpy siempre que se utilicen con métodos optimizados.

